extensions/amp-next-page/0.2/visibility-observer.js

```diff
@@ -0,0 +1,190 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  PositionInViewportEntryDef,
+  PositionObserverFidelity,
+} from '../../../src/service/position-observer/position-observer-worker';
+import {
+  PositionObserver, // eslint-disable-line no-unused-vars
+  installPositionObserverServiceForDoc,
+} from '../../../src/service/position-observer/position-observer-impl';
+import {RelativePositions} from '../../../src/layout-rect';
+import {Services} from '../../../src/services';
+import {devAssert} from '../../../src/log';
+
+/** @enum {number} */
+export const ViewportRelativePos = {
+  INSIDE_VIEWPORT: 0,
+  OUTSIDE_VIEWPORT: 1,
+  LEAVING_VIEWPORT: 2,
+  CONTAINS_VIEWPORT: 3,
+};
+
+export class VisibilityObserverEntry {
+  /**
+   * @param {!VisibilityObserver} observer
+   */
+  constructor(observer) {
+    /** @private {!VisibilityObserver} */
+    this.observer_ = observer;
+    /** @private {?RelativePositions} */
+    this.topSentinelPosition_ = null;
+    /** @private {?RelativePositions} */
+    this.bottomSentinelPosition_ = null;
+    /** @private {!ViewportRelativePos} */
+    this.relativePos_ = ViewportRelativePos.OUTSIDE_VIEWPORT;
+  }
+
+  /**
+   * @param {!Element} element
+   * @param {!Element} parent
+   * @param {!Function} callback
+   */
+  observe(element, parent, callback) {
+    const top = this.observer_.ampdoc.win.document.createElement('div');
```

Comment:
Should be able to simplify to `element.ownerDocument.createElement('div');`. While ownerDocument is typed as nullable, that shouldn't be a problem since it is only null for Documents. I don't think the type check should complain.